File: packages/realtimekit/ios/Plugin/RealtimeKitPlugin.swift
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(RealtimeKitPlugin)
public class RealtimeKitPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "RealtimeKitPlugin"
    public let jsName = "RealtimeKit"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startMeeting", returnType: CAPPluginReturnPromise)
    ]
    public static let tag = "RealtimeKit"

    private let implementation = RealtimeKit()

    @objc func initialize(_ call: CAPPluginCall) {
        if !hasUsageDescription(forKey: "NSBluetoothPeripheralUsageDescription") {
            rejectCall(call, CustomError.privacyDescriptionsMissing)
            return
        }
        if !hasUsageDescription(forKey: "NSBluetoothAlwaysUsageDescription") {
            rejectCall(call, CustomError.privacyDescriptionsMissing)
            return
        }
        if !hasUsageDescription(forKey: "NSCameraUsageDescription") {
            rejectCall(call, CustomError.privacyDescriptionsMissing)
            return
        }
        if !hasUsageDescription(forKey: "NSMicrophoneUsageDescription") {
            rejectCall(call, CustomError.privacyDescriptionsMissing)
            return
        }
        if !hasUsageDescription(forKey: "NSPhotoLibraryUsageDescription") {
            rejectCall(call, CustomError.privacyDescriptionsMissing)
            return
        }
        if !hasUsageDescription(forKey: "UIBackgroundModes") {
            rejectCall(call, CustomError.privacyDescriptionsMissing)
            return
        }

        resolveCall(call)
    }

    @objc func startMeeting(_ call: CAPPluginCall) {
        do {
            let options = StartMeetingOptions(call)
            implementation.startMeeting(options) { error in
                if let error = error {
                    self.rejectCall(call, error)
                    return
                }
                resolveCall(call)
            }
        } catch {
            rejectCall(call, error)
        }
    }

    private func hasUsageDescription(forKey key: String) -> Bool {
        return Bundle.main.object(forInfoDictionaryKey: key) is String
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", RealtimeKitPlugin.tag, "] ", error)
        call.reject(error.localizedDescription)
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }
}

File: packages/realtimekit/android/src/main/java/io/capawesome/capacitorjs/plugins/realtimekit/RealtimeKitPlugin.java
package io.capawesome.capacitorjs.plugins.realtimekit;

import androidx.annotation.NonNull;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.realtimekit.classes.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.realtimekit.classes.options.StartMeetingOptions;

@CapacitorPlugin(name = "RealtimeKit")
public class RealtimeKitPlugin extends Plugin {

    public static final String ERROR_AUTH_TOKEN_MISSING = "authToken must be provided.";
    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";

    private static final String TAG = "RealtimeKit";

    private final RealtimeKit implementation = new RealtimeKit(this);

    @PluginMethod
    public void initialize(PluginCall call) {
        resolveCall(call);
    }

    @PluginMethod
    public void startMeeting(PluginCall call) {
        try {
            StartMeetingOptions options = new StartMeetingOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }

                @Override
                public void success() {
                    resolveCall(call);
                }
            };

            implementation.startMeeting(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        Logger.error(TAG, message, exception);
        call.reject(message);
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }
}
